cranelift-codegen/src/legalizer/mod.rs

```diff
@@ -498,3 +498,33 @@ fn expand_stack_store(
     mflags.set_aligned();
     pos.func.dfg.replace(inst).store(mflags, val, addr, 0);
 }
+
+fn expand_iconst(
+    inst: ir::Inst,
+    func: &mut ir::Function,
+    _cfg: &mut ControlFlowGraph,
+    _isa: &dyn TargetIsa,
+) {
+    let imm: i64 = if let ir::InstructionData::UnaryImm {
+        opcode: ir::Opcode::Iconst,
+        imm,
+    } = &func.dfg[inst]
+    {
+        (*imm).into()
+    } else {
+        panic!("unexpected instruction in expand_iconst");
+    };
+
+    let mut pos = FuncCursor::new(func).at_inst(inst);
+    pos.use_srcloc(inst);
+
+    if pos.func.dfg.ctrl_typevar(inst) == I64 {
```

Comment:
Can you please remove this if, use `pos.func.dfg.ctrl_typevar(inst).half_width().unwrap()` instead of `I32` below and adapt the splitting? That would automatically make `iconst.i128 some_imm_fitting_in_64bits` possible on my `i128` branch. Removing the if would not cause `iconst.i32` to get split unless there is no encoding for it.

Reply:
Nice idea. It would require additional work to compute the right constants for the shift and mask according to the ctrl type, but that's easy.

Reply:
Thinking a bit more about this, while the idea of generalizing this code is tempting, there are 2 things that come to mind:

- the WIP PR #795 doesn't introduce 128-bits immediates (or I missed something), so the instruction format will be different and a different legalization will have to be used.
- having something very general wouldn't bring much, since as far as I know there are no plans to support 8 or 16-bits platforms. So adding one else if when we have i128 support should show simple enough for our needs.

Reply:
> the WIP PR #795 doesn't introduce 128-bits immediates (or I missed something)

correct

> having something very general wouldn't bring much, since as far as I know there are no plans to support 8 or 16-bits platforms. So adding one else if when we have i128 support should show simple enough for our needs.

+1